libdnf/rpm/repo.cpp

```diff
@@ -0,0 +1,2083 @@
+            throw RuntimeError(msg);
+        }
+    }
+    p_impl.reset(new Impl(*this, id, type, std::move(conf), base));
+}
+
+Repo::~Repo() = default;
+
+void Repo::set_callbacks(std::unique_ptr<RepoCB> && callbacks) {
+    p_impl->callbacks = std::move(callbacks);
+}
+
+int Repo::verify_id(const std::string & id) {
+    auto idx = id.find_first_not_of(REPOID_CHARS);
+    return idx == id.npos ? -1 : static_cast<int>(idx);
+}
+
+void Repo::verify() const {
+    if (p_impl->conf->baseurl().empty() &&
+        (p_impl->conf->metalink().empty() || p_impl->conf->metalink().get_value().empty()) &&
+        (p_impl->conf->mirrorlist().empty() || p_impl->conf->mirrorlist().get_value().empty()))
+        throw RuntimeError(fmt::format(_("Repository {} has no mirror or baseurl set."), p_impl->id));
+
+    const auto & type = p_impl->conf->type().get_value();
+    const char * supportedRepoTypes[]{"rpm-md", "rpm", "repomd", "rpmmd", "yum", "YUM"};
+    if (!type.empty()) {
+        for (auto supported : supportedRepoTypes) {
+            if (type == supported) {
+                return;
+            }
+        }
+        throw RuntimeError(
+            fmt::format(_("Repository '{}' has unsupported type: 'type={}', skipping."), p_impl->id, type));
+    }
+}
+
+ConfigRepo * Repo::get_config() noexcept {
+    return p_impl->conf.get();
+}
+
+const std::string & Repo::get_id() const noexcept {
+    return p_impl->id;
+}
+
+void Repo::enable() {
+    p_impl->conf->enabled().set(Option::Priority::RUNTIME, true);
+}
+
+void Repo::disable() {
+    p_impl->conf->enabled().set(Option::Priority::RUNTIME, false);
+}
+
+bool Repo::is_enabled() const {
+    return p_impl->conf->enabled().get_value();
+}
+
+bool Repo::is_local() const {
+    auto & conf = p_impl->conf;
+    if ((!conf->metalink().empty() && !conf->metalink().get_value().empty()) ||
+        (!conf->mirrorlist().empty() && !conf->mirrorlist().get_value().empty())) {
+        return false;
+    }
+    if (!conf->baseurl().get_value().empty() && conf->baseurl().get_value()[0].compare(0, 7, "file://") == 0) {
+        return true;
+    }
+    return false;
+}
+
+bool Repo::load() {
+    return p_impl->load();
+}
+bool Repo::load_cache(bool throw_except) {
+    return p_impl->load_cache(throw_except);
+}
+void Repo::download_metadata(const std::string & destdir) {
+    p_impl->download_metadata(destdir);
+}
+bool Repo::get_use_includes() const {
+    return p_impl->use_includes;
+}
+void Repo::set_use_includes(bool enabled) {
+    p_impl->use_includes = enabled;
+}
+bool Repo::get_load_metadata_other() const {
+    return p_impl->load_metadata_other;
+}
+void Repo::set_load_metadata_other(bool value) {
+    p_impl->load_metadata_other = value;
+}
+int Repo::get_cost() const {
+    return p_impl->conf->cost().get_value();
+}
+int Repo::get_priority() const {
+    return p_impl->conf->priority().get_value();
+}
+std::string Repo::get_comps_fn() {
```

Comment:
Why we have several functions like that represents get_metadata_path with certain string. I suggest that we can move MD_TYPE_MODULES to public part and drop other function. Anyway additional data requeres a general function that will allow to use any string. What do you think?

Reply:
But I see that there is an additional logic - MD_TYPE_GROUP_GZ, MD_TYPE_GROUP.  A similar problem is with zchunk data.

Reply:
I will remove `Repo::get_comps_fn()` and `Repo::get_modules_fn()` methods.